components/director/internal/domain/application/service.go

```diff
@@ -712,6 +715,60 @@ func (s *service) DeleteLabel(ctx context.Context, applicationID string, key str
 	return nil
 }
 
+// Merge Merges properties from Source Application into Destination Application, provided that the Destination's
+// Application does not have a value set for a given property. Then the Source Application is being deleted.
+func (s *service) Merge(ctx context.Context, destID, srcID string) (*model.Application, error) {
+	appTenant, err := tenant.LoadFromContext(ctx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while loading tenant from context")
+	}
+
+	destApp, err := s.Get(ctx, destID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while getting destination application")
+	}
+
+	srcApp, err := s.Get(ctx, srcID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while getting source application")
+	}
+
+	srcBaseURL := strings.TrimSuffix(str.PtrStrToStr(srcApp.BaseURL), urlSuffixToBeTrimmed)
+	destBaseURL := strings.TrimSuffix(str.PtrStrToStr(destApp.BaseURL), urlSuffixToBeTrimmed)
+	if len(srcBaseURL) == 0 || len(destBaseURL) == 0 || srcBaseURL != destBaseURL {
+		return nil, errors.Errorf("BaseURL for applications %s and %s are not the same. Destination app BaseURL: %s. Source app BaseURL: %s", destID, srcID, destBaseURL, srcBaseURL)
+	}
+
+	srcTemplateID := str.PtrStrToStr(srcApp.ApplicationTemplateID)
+	destTemplateID := str.PtrStrToStr(destApp.ApplicationTemplateID)
+	if len(srcTemplateID) == 0 || len(destTemplateID) == 0 || srcTemplateID != destTemplateID {
+		return nil, errors.Errorf("Application templates are not the same. Destination app template: %s. Source app template: %s", destTemplateID, srcTemplateID)
+	}
+
+	if srcApp.Status == nil {
+		return nil, errors.Errorf("Could not determine status of source application with id %s", srcID)
+	}
+
+	if srcApp.Status.Condition == model.ApplicationStatusConditionConnected {
+		return nil, errors.Errorf("Cannot merge application with id %s, because it is in a %s status", srcID, model.ApplicationStatusConditionConnected)
+	}
+
+	if err := mergo.Merge(destApp, *srcApp); err != nil {
```

Comment:
Just realized we'd be having a nasty bug if all we did was merge the fields and forget the labels.

These are the types of labels associated with a autodiscovered app:
<img width="585" alt="image" src="https://user-images.githubusercontent.com/15074116/165306487-4a482520-7f74-4f0b-8a88-d4ead587c4d8.png">


These are the types of labels associated with a self-reg app:
<img width="351" alt="image" src="https://user-images.githubusercontent.com/15074116/165306520-8d8a0efa-fe9d-4fb1-9026-34a3b9172fb0.png">

Ideally this should be fixed in the same PR, but since the API is not open for productive use we can think about splitting it or opening a bug. Let's discuss this..

Reply:
Also the E2E test should ideally cover this labels scenario as well.